### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Yoel-Perman-Brilliant**

### CODE REVIEW

File: src/main/java/com/spikes2212/command/drivetrains/TankDrivetrain.java [hunk 1]
```diff
@@ -51,42 +50,62 @@ public void tankDrive(double speedLeft, double speedRight) {
...
      */
     public void tankDrive(double speedLeft, double speedRight, boolean squareInputs) {
         drive.tankDrive(speedLeft, speedRight, squareInputs);
     }
 
     /**
-     * Moves the drivetrain with the given forward and angular speed.
+     * Moves both sides of this drivetrain by the given voltages for each side.
+     *
+     * @param leftVoltage  the voltage to the left side (-1 to 1). Positive values move this side forward
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/drivetrains/TankDrivetrain.java:64]: -12 to 12

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ */
+public class DriveTankWithVoltages extends CommandBase {
+
+    protected final Supplier<Double> leftVoltageSupplier;
+    protected final Supplier<Double> rightVoltageSupplier;
+    protected final TankDrivetrain drivetrain;
+    protected final Supplier<Boolean> isFinished;
+
+    /**
+     * This constructs a new {@link DriveTankWithVoltages} command that moves the given {@link TankDrivetrain} according to voltage
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:22]: to the given voltage

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    public DriveTankWithVoltages(TankDrivetrain drivetrain, Supplier<Double> leftVoltageSupplier,
+                                 Supplier<Double> rightVoltageSupplier, Supplier<Boolean> isFinished) {
+        this.drivetrain = drivetrain;
+        this.leftVoltageSupplier = leftVoltageSupplier;
+        this.rightVoltageSupplier = rightVoltageSupplier;
+        this.isFinished = isFinished;
+    }
+
+    /**
+     * This constructs a new {@link DriveTankWithVoltages} command that moves the given {@link TankDrivetrain} according to voltage
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:40]: ditto

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+     * @param leftVoltageSupplier  the double {@link Supplier} supplying the voltage for the left side (-12 to 12)
+     * @param rightVoltageSupplier the double {@link Supplier} supplying the voltage for the right side (-12 to 12)
+     */
+    public DriveTankWithVoltages(TankDrivetrain drivetrain, Supplier<Double> leftVoltageSupplier,
+                                 Supplier<Double> rightVoltageSupplier) {
+        this(drivetrain, leftVoltageSupplier, rightVoltageSupplier, () -> false);
+    }
+
+    /**
+     * This constructs a new {@link DriveTankWithVoltages} command that moves the given {@link TankDrivetrain} according to voltage
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:54]: ditto

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+     * @param rightVoltage the double {@link Supplier} supplying the voltage for the right side (-12 to 12)
+     * @param isFinished   when to finish the command
+     */
+    public DriveTankWithVoltages(TankDrivetrain drivetrain, double leftVoltage, double rightVoltage,
+                                 Supplier<Boolean> isFinished) {
+        this(drivetrain, () -> leftVoltage, () -> rightVoltage, isFinished);
+    }
+
+    /**
+     * This constructs a new {@link DriveTankWithVoltages} command that moves the given {@link TankDrivetrain} according to voltage
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:69]: ditto

File: src/main/java/com/spikes2212/util/Limelight.java [hunk 6]
```diff
@@ -206,6 +217,6 @@ public void setCamMode(CamMode mode) {
      */
     public void setLedMode(LedMode mode) {
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/util/Limelight.java:218]: docs

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+     * values for left and right sides. <br>
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:70]: te

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 8]
```diff
@@ -16,12 +20,24 @@ public interface Namespace {
...
-    Namespace addChild(String name);
+    ChildNamespace addChild(String name);
 
     void putData(String key, Sendable value);
 
+    default void putCommand(String key, Command command, boolean ignoringDisable) {
+        this.putData(key, command.ignoringDisable(ignoringDisable));
+    }
+
+    default void putCommand(String key, Command command) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:31]: document all and specify that it runs on disable

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 9]
```diff
@@ -50,5 +66,24 @@ default void putBoolean(String key, boolean value) {
 
     boolean getBoolean(String key);
 
+    default PIDSettings addPIDNamespace(String name, PIDSettings initialPIDSettings) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:69]: test and document

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 10]
```diff
@@ -50,5 +66,24 @@ default void putBoolean(String key, boolean value) {
...
+        ChildNamespace child = this.addChild(name + " pid");
+        Supplier<Double> kP = child.addConstantDouble("kP " + name, initialPIDSettings.getkP());
+        Supplier<Double> kI = child.addConstantDouble("kI " + name, initialPIDSettings.getkI());
+        Supplier<Double> kD = child.addConstantDouble("kD " + name, initialPIDSettings.getkD());
+        Supplier<Double> tolerance = child.addConstantDouble(name + " tolerance", initialPIDSettings.getTolerance());
+        Supplier<Double> waitTime = child.addConstantDouble(name + " wait time", initialPIDSettings.getWaitTime());
+        return new PIDSettings(kP, kI, kD, tolerance, waitTime);
+    }
+
+    default FeedForwardSettings addFeedForwardNamespace(String name, FeedForwardSettings initialFeedForwardSettings) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:79]: ditto

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * Sets a certain range of LEDs to a specific color.
+     *
+     * @param start the first LED in the range
+     * @param end   the final LED in the range
+     * @param color the desired {@link Color}
+     */
+    public void setColorInRange(int start, int end, Color color) {
+        setColorInRange(color.getRed(), color.getGreen(), color.getBlue(), start, end);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:81]: wrong order

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * Sets the LED strip to a specific color.
+     *
+     * @param red   the red value from 0 to 255
+     * @param green the green value from 0 to 255
+     * @param blue  the blue value from 0 to 255
+     */
+    public void setStripColor(int red, int green, int blue) {
+        setColorInRange(red, green, blue, 0, ledBuffer.getLength());
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:39]: wrong order

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * Sets a specific LED to a specific color.
+     *
+     * @param index the index of the LED
+     * @param red   the red value from 0 to 255
+     * @param green the green value from 0 to 255
+     * @param blue  the blue value from 0 to 255
+     */
+    public void setColorAt(int index, int red, int green, int blue) {
+        ledBuffer.setRGB(index, red, green, blue);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:93]: check order

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltages.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+public class DriveArcadeWithVoltages extends CommandBase {
+
+    protected final TankDrivetrain tankDrivetrain;
+    protected final Supplier<Double> moveValueSupplier;
+    protected final Supplier<Double> rotateValueSupplier;
+    protected final Supplier<Boolean> isFinished;
+
+    /**
+     * Constructs a new {@link DriveArcade} command that moves the given
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltages.java:23]: DriveArcadeWithVoltages!

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 15]
```diff
@@ -10,45 +15,217 @@
  */
 public interface Namespace {
 
+    /**
+     * Adds a double value to the namespace.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:19]: supplier

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 16]
```diff
@@ -10,45 +15,217 @@
...
+     * Can only be changed via the {@link Shuffleboard}, or any other {@link NetworkTable}s UI.
+     *
+     * @param name  the key that will be given to the value
+     * @param value the initial value to be added
+     * @return a {@link Supplier} with the value set using the {@link NetworkTable}s
+     */
     Supplier<Double> addConstantDouble(String name, double value);
 
+    /**
+     * Adds an integer value to the namespace.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:29]: ditto

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 17]
```diff
@@ -10,45 +15,217 @@
...
+     * Can only be changed via the {@link Shuffleboard}, or any other {@link NetworkTable}s UI.
+     *
+     * @param name  the key that will be given to the value
+     * @param value the initial value to be added
+     * @return a {@link Supplier} with the value set using the {@link NetworkTable}s
+     */
     Supplier<Integer> addConstantInt(String name, int value);
 
+    /**
+     * Adds a string value to the namespace.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:39]: ditto

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 18]
```diff
@@ -10,45 +15,217 @@
  */
 public interface Namespace {
 
+    /**
+     * Adds a double value to the namespace.
+     * Can only be changed via the {@link Shuffleboard}, or any other {@link NetworkTable}s UI.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:20]: whose value can...

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 19]
```diff
@@ -10,45 +15,217 @@
...
+        Supplier<Double> kD = child.addConstantDouble("kD " + name, initialPIDSettings.getkD());
+        Supplier<Double> tolerance = child.addConstantDouble(name + " tolerance", initialPIDSettings.getTolerance());
+        Supplier<Double> waitTime = child.addConstantDouble(name + " wait time", initialPIDSettings.getWaitTime());
+        return new PIDSettings(kP, kI, kD, tolerance, waitTime);
+    }
+
+    /**
+     * Adds a set of {@link FeedForwardSettings} values to the namespace to a designated {@link ChildNamespace}.
+     *
+     * @param name               the name to be given to the settings
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:214]: and the child namespace

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 20]
```diff
@@ -0,0 +1,126 @@
+package com.spikes2212.util;
+
+import com.spikes2212.command.DashboardedSubsystem;
+import edu.wpi.first.wpilibj.AddressableLED;
+import edu.wpi.first.wpilibj.AddressableLEDBuffer;
+
+import java.awt.*;
+
+/**
+ * A class that handles the communication between an LED strip and the code.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:10]: add a note about testing

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+     *
+     * @param color the desired {@link Color}
+     * @param index the index of the LED
+     */
+    public void setColorAt(int index, Color color) {
+        ledBuffer.setRGB(index, color.getRed(), color.getGreen(), color.getBlue());
+    }
+
+    /**
+     * Takes the buffer's data and applies it to the LED strip periodically.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:107]: should be called periodically

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 22]
```diff
@@ -10,45 +15,217 @@
...
     void putData(String key, Sendable value);
 
+    /**
+     * Adds a command to the namespace.
+     *
+     * @param key             the key that will be given to the value
+     * @param command         the command to be added
+     * @param ignoringDisable whether the command should be executable when the robot is disabled
+     */
+    default void putCommand(String key, Command command, boolean ignoringDisable) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:71]: runOnDisable

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 23]
```diff
@@ -10,45 +15,217 @@
...
+    /**
+     * Adds a {@link ChildNamespace} to this namespace.
+     *
+     * @param name the name given to the child namespace's name
+     * @return the new {@link ChildNamespace} that is created
+     */
+    ChildNamespace addChild(String name);
+
+    /**
+     * Adds {@link Sendable} data to the namespace.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:57]: Wrong. -5

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 24]
```diff
@@ -10,45 +15,217 @@
...
+     * @param value the value to be added
+     */
     void putData(String key, Sendable value);
 
+    /**
+     * Adds a command to the namespace.
+     *
+     * @param key             the key that will be given to the value
+     * @param command         the command to be added
+     * @param runOnDisable whether the command should be executable when the robot is disabled
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:69]: this param description is misaligned

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 25]
```diff
@@ -10,45 +15,217 @@
...
+     *
+     * @param key     the key that will be given to the value
+     * @param command the command to be added
+     */
+    default void putCommand(String key, Command command) {
+        this.putCommand(key, command, true);
+    }
+
+    /**
+     * Adds a runnable value to the namespace that can be run as a {@link InstantCommand}.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:86]: 'an InstantCommand' -1

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 26]
```diff
@@ -10,45 +15,217 @@
...
+     * @param key          the key that will be given to the value
+     * @param runnable     the runnable value to be added
+     * @param runOnDisable whether the instant command should be executable when the robot is disabled
+     */
+    default void putRunnable(String key, Runnable runnable, boolean runOnDisable) {
+        this.putData(key, new InstantCommand(runnable).ignoringDisable(runOnDisable));
+    }
+
+    /**
+     * Adds a runnable value that can be run as a {@link InstantCommand} and can run on disable.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:97]: 'an InstantCommand' -1

Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:97]: this wording is inconsistent with other places where you describe the ability of a command to run when the robot is disabled -1

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 27]
```diff
@@ -10,45 +15,217 @@
...
+        Supplier<Double> tolerance = child.addConstantDouble(name + " tolerance", initialPIDSettings.getTolerance());
+        Supplier<Double> waitTime = child.addConstantDouble(name + " wait time", initialPIDSettings.getWaitTime());
+        return new PIDSettings(kP, kI, kD, tolerance, waitTime);
+    }
+
+    /**
+     * Adds a set of {@link FeedForwardSettings} values to the namespace to a designated {@link ChildNamespace}.
+     *
+     * @param name                       the name to be given to the settings and the child namespace
+     * @param initialFeedForwardSettings the feed forward settings to be added
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:215]: specify that these are the initial values only, similarly to how it was done with the PIDSettings

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 28]
```diff
@@ -10,45 +15,217 @@
...
+     * @return feed forward settings with the most recent value from the network tables
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:216]: make this consistent with the wording of the previous method's documentation -1

File: src/main/java/com/spikes2212/dashboard/SpikesLogger.java [hunk 29]
```diff
@@ -0,0 +1,60 @@
+package com.spikes2212.dashboard;
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/SpikesLogger.java:1]: add an empty line after this

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 30]
```diff
@@ -0,0 +1,128 @@
+package com.spikes2212.util;
+
+import com.spikes2212.command.DashboardedSubsystem;
+import edu.wpi.first.wpilibj.AddressableLED;
+import edu.wpi.first.wpilibj.AddressableLEDBuffer;
+
+import java.awt.*;
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:7]: why import the whole package?

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+ *
+ * @author Camellia Lami
+ * @see DashboardedSubsystem
+ */
+public class AddressableLEDWrapper {
+
+    /**
+     * The controlled LED strip.
+     */
+    private final AddressableLED led;
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:21]: add an empty line after this one

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 32]
```diff
@@ -0,0 +1,128 @@
...
+     */
+    public void setColorAt(int index, int red, int green, int blue) {
+        ledBuffer.setRGB(index, red, green, blue);
+    }
+
+    /**
+     * Sets a specific LED to a specific color.
+     *
+     * @param color the desired {@link Color}
+     * @param index the index of the LED
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:101]: reorder these description to match the order of the parameters

File: src/main/java/com/spikes2212/dashboard/Namespace.java (hunk 28 above)
Comment (Spikes-2212) [reply src/main/java/com/spikes2212/dashboard/Namespace.java:216]: it literally is????

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]

File: src/main/java/com/spikes2212/command/drivetrains/TankDrivetrain.java [hunk 33]
```diff
@@ -31,62 +32,80 @@ public TankDrivetrain(String namespaceName, MotorController left, MotorControlle
         rightController.setInverted(true);
         drive = new DifferentialDrive(leftController, rightController);
     }
-    
+
     public TankDrivetrain(MotorController left, MotorController right) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/TankDrivetrain.java:36]: add a constructor which takes 4 motor controllers; 2 on each side

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 34]
```diff
@@ -0,0 +1,129 @@
...
+import java.awt.Color;
+
+/**
+ * A class that handles the communication between an LED strip and the code.
+ * This class still requires further testing and an update will be made as soon as testing is finished.
+ *
+ * @author Camellia Lami
+ * @see DashboardedSubsystem
+ */
+public class AddressableLEDWrapper {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:16]: add a few functions which make the leds cool (such as advanceLeds)

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]

**Review by Spikes-2212 — APPROVED** [review]

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 35]
```diff
@@ -10,45 +15,217 @@
...
 
+    /**
+     * Gets a Number value from the namespace.
+     *
+     * @param key the key of the value
+     * @return the desired value
+     */
     double getNumber(String key);
 
+
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:166]: delete

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 36]
```diff
@@ -10,45 +15,217 @@
...
+    /**
+     * Adds a Boolean {@link Supplier} to the namespace.
+     *
+     * @param key   the key that will be given to the value
+     * @param value the value to be added
+     */
     void putBoolean(String key, Supplier<Boolean> value);
 
+
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:175]: delete

File: src/main/java/com/spikes2212/dashboard/Namespace.java [hunk 37]
```diff
@@ -10,45 +15,217 @@
...
+     * @return the desired value
+     */
     boolean getBoolean(String key);
 
+    /**
+     * Adds a set of {@link PIDSettings} values to a designated {@link ChildNamespace}.
+     *
+     * @param name               the name to be given to the settings and the child namespace
+     * @param initialPIDSettings the initial values for the PID settings to be added
+     * @return PID settings with the values from the network tables
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/Namespace.java:199]: uncapitalize

File: src/main/java/com/spikes2212/dashboard/SpikesLogger.java [hunk 38]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Creates a SpikesLogger instance with custom name and key for the output location.
+     */
+    public SpikesLogger(String name, String key) {
+        super(name);
+        this.key = key;
+    }
+
+    /**
+     * Creates a default SpikesLogger instance with custom key (name = "SpikesLogger").
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/SpikesLogger.java:29]: a

File: src/main/java/com/spikes2212/dashboard/SpikesLogger.java [hunk 39]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Logs the provided output to the NetworkTables and the SpikesLogger app.
+     *
+     * @param output the data to be logged
+     */
+    public <T> void log(T output) {
+        putString(key, output == null ? "null" : output.toString());
+    }
+
+
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/dashboard/SpikesLogger.java:51]: delete

File: src/main/java/com/spikes2212/util/AddressableLEDWrapper.java [hunk 40]
```diff
@@ -0,0 +1,129 @@
...
+     * Sets a certain range of LEDs to a specific color.
+     *
+     * @param start the first LED in the range
+     * @param end   the final LED in the range
+     * @param red   the red value from 0 to 255
+     * @param green the green value from 0 to 255
+     * @param blue  the blue value from 0 to 255
+     */
+    public void setColorInRange(int start, int end, int red, int green, int blue) {
+        for (int i = start; i < end; i++) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/AddressableLEDWrapper.java:70]: <=

File: src/main/java/com/spikes2212/util/Limelight.java [hunk 41]
```diff
@@ -58,18 +59,28 @@ public Limelight() {
     }
 
     /**
-     * Retrieves an entry from the Limelight NetworkTable.
+     * Retrieves an entry from the Limelight's NetworkTable.
      *
      * @param key key for entry
      * @return the value of the given entry
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/Limelight.java:65]: given key's entry

File: src/main/java/com/spikes2212/dashboard/SpikesLogger.java [hunk 42]
```diff
@@ -0,0 +1,60 @@
+package com.spikes2212.dashboard;
+
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+
+/**
+ * A logger class meant to be used with the SpikesLogger desktop app to log values from the robot to a computer in real-time.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/dashboard/SpikesLogger.java]: add link

**Review by Spikes-2212 — APPROVED** [review]

**Review by TurtleMinecraft — APPROVED** [review]

**Pull request merged**